### ISSUE TITLE
Devise session reset on subsequent request

### ISSUE THREAD

**Issue by saveav** [plataformatec/devise#3623]
I used:

Devise (3.5.1)
Omniauth (1.2.2)
Warden (1.2.3)

On a trial project, a new one using Mongoid as the ORM. Sessions controller are created anew, and used, instead of the Devise's standard, and changed into:

```ruby
class UserDevise::SessionsController < Devise::SessionsController
  def new
    super
  end

  def create
    puts "AUTH OPTIONS"
    puts auth_options
    self.resource = warden.authenticate!(auth_options)
    puts "RESOURCE"
    puts self.resource.inspect
    set_flash_message(:notice, :signed_in) if is_flashing_format?
    sign_in(resource_name, resource)
    yield resource if block_given?

    session[:whatever] = "ADAM PAHLEVI"

    puts session.keys

    flash.delete(:notice)

    if current_user.data.prereg
      if params[:original_url]
        redirect_to params[:original_url]
      else
        redirect_to main_app.root_path
      end
    else
      # prereg_path
      redirect_to main_app.root_path
    end
  end

  def destroy
    signed_out = (Devise.sign_out_all_scopes ? sign_out : sign_out(resource_name))
    super
    flash.delete(:notice)
  end
end
```

My session (`whatever`) is persisted, but Warden's session is reset on subsequent request. I don't know why.

**Comment by josevalim** [comment]
Are you doing sign up with ajax or something like that? In any case, it is very hard to find out what is wrong with such a code snippets, so  Can you please provide a sample application that reproduces the error?

**Comment by saveav** [comment]
No Ajax. I am new to Git, should I set up a new repo to this? I used git mainly only to push something.

**Comment by saveav** [comment]
I already undergone several test in different rails version. And, the strange thing is not in the devise itself, but rather in MongoDB, or should i say... mongoid?

If a mongo db is shared by two different app... what happen is that, you can login (under same devise configuration) but the session won't persist.

I don't know if this is a mongo/rails/devise problem...

**Comment by saveav** [comment]
I confirm that there's problem with sharing mongodb database for devise's user login.

I used an entirely new project, new, as in "rails new." First, with a separate database, everything is OK. Then, I changed the db to use the same mongoid db, and then problem begin to occur: that is, session not persisted.

Strange...

**Comment by scottlowe** [comment]
@saveav Are you still seeing this issue? I am using Devise and Mongoid and may have a similar problem - I seem to be losing authenticated state within Warden. Trying to work out if this is related.

**Comment by josevalim** [comment]
Which mongoid version are you using? It was a mongoid bug which should hopefully be fixed.

**Comment by scottlowe** [comment]
Using Mongoid 4.0.2. So this should be fixed in Devise 3.5.2?
